packages/noya-colorpicker/src/components/GradientSlider.tsx

```diff
@@ -0,0 +1,204 @@
+const getRelativePosition = (
+  node: HTMLDivElement,
+  event: MouseEvent | TouchEvent,
+): Interaction => {
+  const rect = node.getBoundingClientRect();
+
+  // Get user's pointer position from `touches` array if it's a `TouchEvent`
+  const pointer = isTouch(event) ? event.touches[0] : (event as MouseEvent);
+
+  return {
+    left: clamp(
+      (pointer.pageX - (rect.left + window.pageXOffset)) / rect.width,
+    ),
+    top: clamp((pointer.pageY - (rect.top + window.pageYOffset)) / rect.height),
+  };
+};
+
+interface Props {
+  onMove: (interaction: Interaction) => void;
+  onMoveChange: (moving: boolean) => void;
+  onClick: (interaction: Interaction) => void;
+  onKey: (offset: Interaction) => void;
+  isOnPoint: (interaction: Interaction) => boolean;
+  children: React.ReactNode;
+}
+
+const InteractiveBase = ({
+  onMove,
+  onKey,
+  onClick,
+  onMoveChange,
+  isOnPoint,
+  ...rest
+}: Props) => {
+  const container = useRef<HTMLDivElement>(null);
+  const hasTouched = useRef(false);
+  const isDragging = useRef(false);
+  const onMoveCallback = useEventCallback(onMove);
+  const onKeyCallback = useEventCallback(onKey);
+  const onClickCallback = useEventCallback(onClick);
+
+  // Prevent mobile browsers from handling mouse events (conflicting with touch ones).
+  // If we detected a touch interaction before, we prefer reacting to touch events only.
+  const isValid = (event: MouseEvent | TouchEvent): boolean => {
+    if (hasTouched.current && !isTouch(event)) return false;
+    if (!hasTouched.current) hasTouched.current = isTouch(event);
+    return true;
+  };
+
+  const handleMove = useCallback(
+    (event: MouseEvent | TouchEvent) => {
+      if (!isDragging.current || !container.current) return;
+
+      event.preventDefault();
+      event.stopPropagation();
+      event.stopImmediatePropagation();
+
+      onMoveCallback(getRelativePosition(container.current, event));
+
+      // // If user moves the pointer outside of the window or iframe bounds and release it there,
+      // // `mouseup`/`touchend` won't be fired. In order to stop the picker from following the cursor
+      // // after the user has moved the mouse/finger back to the document, we check `event.buttons`
+      // // and `event.touches`. It allows us to detect that the user is just moving his pointer
+      // // without pressing it down
+      // const isDown = isTouch(event)
+      //   ? event.touches.length > 0
+      //   : event.buttons > 0;
+
+      // if (isDown && container.current) {
+      //   onMoveCallback(getRelativePosition(container.current, event));
+      // } else {
+      //   isDragging.current = false;
+      // }
+    },
+    [onMoveCallback],
+  );
+
+  const handleMoveStart = useCallback(
+    ({ nativeEvent: event }: React.MouseEvent | React.TouchEvent) => {
+      event.preventDefault();
+
+      if (!isValid(event)) return;
+
+      if (!isOnPoint(getRelativePosition(container.current!, event))) {
```

Comment:
We could probably use `event.target` to determine if we're clicking a pointer. E.g. we could add a `className="pointer"` to the pointer elements, and then:

```ts
if (event.target instanceof HTMLElement && event.target.classList.contains('pointer')) {
  onClickCallback(getRelativePosition(container.current!, event));
}
```